### DISCUSSION TITLE
Why Stick to Vanilla Residual Paths in NanoChat?

### DISCUSSION THREAD

**Discussion by Mithun-033** [karpathy/nanochat#712] [Q&A]
I noticed the model still uses standard residual connections, even though they’re known to hurt gradient flow to deeper layers and kind of force those layers to make bigger updates.
Is it because other options are computationally intensive or does it simply not matter at this small scale ?

**Comment by Mithun-033** [comment]
adding on to this, i tested vanilla residual vs scaled residual similar to deepseek implementation on a ~117m model with only 16 layers, got surprising loss diff after just 1bil tokens , 
         
```python
class Block(nn.Module):
    def __init__(self,config):
        super().__init__()
        self.PreNorm1=nn.LayerNorm(config.n_embed,eps=1e-5)
        self.attention=MultiHeadAttention(config)
        self.PreNorm2=nn.LayerNorm(config.n_embed,eps=1e-5)
        self.FeedForwardLayer=MLP(config)
        self.res_scale=1/math.sqrt(2*config.n_layer)

    def forward(self,x):
        x=x+self.res_scale*self.attention(self.PreNorm1(x))
        x=x+self.res_scale*self.FeedForwardLayer(self.PreNorm2(x))
        return x
``` 
          could work the same for nanochat?

**Reply by Mithun-033** [reply]
jus saw nanochat uses a better scaled residual
